File: pc1/sensores.py
# sensores.py - simulación de sensores de tráfico - pc1
# 
# este script simula los 3 tipos de sensores de tráfico que hay 
# en la ciudad:
#   1. cámaras de tráfico -> miden longitud de cola (q) y velocidad (vp)
#   2. espiras inductivas -> cuentan vehículos que pasan (cv)
#   3. sensores gps -> miden densidad (d) y velocidad promedio (vp)
# 
# la ciudad es una cuadrícula de 5x5 (filas a-e, columnas 1-5).
# cada sensor genera un evento json cada cierto tiempo y lo publica
# al broker usando pub/sub de zeromq.
# 
# autores: miguel angel acuña, juan david acuña, y samuel felipe manrique - sistemas distribuidos 2026-10

import zmq
import json
import time
import random
import threading
from datetime import datetime, timezone

# ============================================================
# CONFIGURACIÓN - Cambiar la IP del broker (PC1)
# ============================================================
BROKER_IP = "10.43.98.198"
PUERTO_BROKER = 5555  # Puerto XSUB del broker

INTERVALO = 10  # Segundos entre cada evento generado

# Cuadrícula de la ciudad 5x5
FILAS = ["A", "B", "C", "D", "E"]
COLUMNAS = [1, 2, 3, 4, 5]

# Intersecciones que vamos a usar para la demo
INTERSECCIONES = [("A", 1), ("B", 3), ("C", 5), ("D", 2), ("E", 4)]

# TODO: Implementar CurveZMQ para la entrega final


def timestamp_ahora():
    # devuelve la fecha y hora actual en formato iso 8601.
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def generar_evento_camara(fila, col):
    # genera un evento de cámara de tráfico.
    # mide la longitud de cola (q) y velocidad promedio (vp).
    evento = {
        "sensor_id": f"CAM-{fila}{col}",
        "tipo_sensor": "camara",
        "interseccion": f"INT-{fila}{col}",
        "volumen": random.randint(0, 30),          # Q - vehículos en cola
        "velocidad_promedio": round(random.uniform(0, 50), 1),  # Vp - km/h
        "timestamp": timestamp_ahora()
    }
    return evento


def generar_evento_espira(fila, col):
    # genera un evento de espira inductiva.
    # cuenta cuántos vehículos pasaron en un intervalo de 
    # 30 segundos.
    evento = {
        "sensor_id": f"ESP-{fila}{col}",
        "tipo_sensor": "espira_inductiva",
        "interseccion": f"INT-{fila}{col}",
        "vehiculos_contados": random.randint(0, 40),
        "intervalo_segundos": 30,
        "timestamp_inicio": timestamp_ahora(),
        "timestamp_fin": timestamp_ahora()
    }
    return evento


def generar_evento_gps(fila, col):
    # genera un evento del sensor gps.
    # mide velocidad promedio (vp) y densidad de tráfico (d).
    # el nivel de congestión se determina por la velocidad.
    velocidad = round(random.uniform(0, 60), 1)

    # Determino el nivel de congestión según la velocidad
    if velocidad < 10:
        nivel = "ALTA"
        densidad = random.randint(40, 80)
    elif velocidad <= 40:
        nivel = "NORMAL"
        densidad = random.randint(15, 45)
    else:
        nivel = "BAJA"
        densidad = random.randint(1, 20)

    evento = {
        "sensor_id": f"GPS-{fila}{col}",
        "tipo_sensor": "gps",
        "interseccion": f"INT-{fila}{col}",
        "nivel_congestion": nivel,
        "velocidad_promedio": velocidad,
        "densidad": densidad,
        "timestamp": timestamp_ahora()
    }
    return evento


def ejecutar_sensor(tipo, fila, col, contexto):
    # función que ejecuta un sensor individual en un hilo.
    # cada sensor genera eventos periódicamente y los publica al broker.
    # 
    # el mensaje se envía como: "topico {json}"
    # donde topico es: camara, espira, o gps
    # Creo el socket PUB y me conecto al broker
    socket = contexto.socket(zmq.PUB)
    socket.connect(f"tcp://{BROKER_IP}:{PUERTO_BROKER}")

    nombre = f"{tipo.upper()[:3]}-{fila}{col}"
    print(f"[SENSOR] {nombre} conectado al broker")

    # Espero un poco para que la conexión se estabilice
    time.sleep(1)

    while True:
        # Genero el evento según el tipo de sensor
        if tipo == "camara":
            evento = generar_evento_camara(fila, col)
            topico = "camara"
        elif tipo == "espira":
            evento = generar_evento_espira(fila, col)
            topico = "espira"
        else:
            evento = generar_evento_gps(fila, col)
            topico = "gps"

        # Marshalling: convierto el diccionario a JSON (representación externa)
        mensaje_json = json.dumps(evento)

        # Envío el mensaje al broker: "topico JSON"
        socket.send_string(f"{topico} {mensaje_json}")

        print(f"[SENSOR] {nombre} -> {topico} | {evento['interseccion']}")

        # Espero el intervalo antes de generar otro evento
        time.sleep(INTERVALO)


def main():
    print("=" * 60)
    print("  SENSORES DE TRÁFICO - PC1")
    print("=" * 60)
    print(f"  Broker: tcp://{BROKER_IP}:{PUERTO_BROKER}")
    print(f"  Intervalo: {INTERVALO} segundos")
    print(f"  Intersecciones: {len(INTERSECCIONES)}")
    print("=" * 60)

    contexto = zmq.Context()
    hilos = []

    # Creo un hilo por cada sensor en cada intersección
    for fila, col in INTERSECCIONES:
        for tipo in ["camara", "espira", "gps"]:
            t = threading.Thread(
                target=ejecutar_sensor,
                args=(tipo, fila, col, contexto),
                daemon=True
            )
            t.start()
            hilos.append(t)

    total = len(INTERSECCIONES) * 3
    print(f"\n[SENSORES] {total} sensores iniciados")
    print("[SENSORES] Presione Ctrl+C para detener\n")

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\n[SENSORES] Cerrando sensores...")
        print("[SENSORES] Listo.")


if __name__ == "__main__":
    main()

File: java/pc1/src/main/java/Sensores.java
// sensores.java - simulacion de sensores de trafico - pc1
//
// este script simula los 3 tipos de sensores de trafico que hay
// en la ciudad:
//   1. camaras de trafico -> miden longitud de cola q y velocidad vp
//   2. espiras inductivas -> cuentan vehiculos que pasan cv
//   3. sensores gps -> miden densidad d y velocidad promedio vp
//
// la ciudad es una cuadricula de 5x5 filas a-e columnas 1-5
// cada interseccion tiene sensores separados para la carrera y la calle
// cada sensor genera un evento json cada cierto tiempo y lo publica al broker
//
// soporta simulacion de fallos por linea de comandos:
//   -ruptura: desactiva el 20% de los sensores
//   -omision x: omite el x% de los envios de red
//   -temporizacion y: retarda el envio y milisegundos
//
// autores: miguel angel acuna, juan david acuna, y samuel felipe manrique - sistemas distribuidos 2026-10

import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;
import org.json.JSONObject;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Random;

public class Sensores {

    // configuracion - cambiar la ip del broker pc1
    static String BROKER_IP = "10.43.98.198";
    static int PUERTO_BROKER = 5555;  // puerto sub del broker

    static int INTERVALO = 10;  // segundos entre cada evento generado

    // intersecciones que vamos a usar para la demo
    static String[][] INTERSECCIONES = {
        {"A", "1"}, {"B", "3"}, {"C", "5"}, {"D", "2"}, {"E", "4"}
    };

    // parametros de simulacion de fallos
    static boolean simularRuptura = false;
    static int tasaOmision = 0;       // porcentaje de 0 a 100
    static int retardoMili = 0;       // milisegundos de retraso

    static Random random = new Random();

    // devuelve la fecha y hora actual en formato iso 8601
    static String timestampAhora() {
        return DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
                .withZone(ZoneOffset.UTC)
                .format(Instant.now());
    }

    // genera un evento de camara de trafico
    static JSONObject generarEventoCamara(String fila, int col, String via) {
        JSONObject evento = new JSONObject();
        evento.put("sensor_id", "CAM-" + fila + col + "-" + via);
        evento.put("tipo_sensor", "camara");
        evento.put("interseccion", "INT-" + fila + col);
        evento.put("via", via);
        
        // simular trafico congestionado de vez en cuando el resto normal
        boolean esNormal = random.nextInt(100) < 75;
        int q;
        double vp;
        
        if (esNormal) {
            q = random.nextInt(5); // 0-4 vehiculos en cola
            vp = 36.0 + random.nextDouble() * 14.0; // 36-50 km/h
        } else {
            q = 12 + random.nextInt(15); // 12-26 vehiculos
            vp = 5.0 + random.nextDouble() * 15.0; // 5-20 km/h
        }
        
        evento.put("volumen", q);
        evento.put("velocidad_promedio", Math.round(vp * 10.0) / 10.0);
        evento.put("timestamp", timestampAhora());
        return evento;
    }

    // genera un evento de espira inductiva
    static JSONObject generarEventoEspira(String fila, int col, String via) {
        JSONObject evento = new JSONObject();
        evento.put("sensor_id", "ESP-" + fila + col + "-" + via);
        evento.put("tipo_sensor", "espira_inductiva");
        evento.put("interseccion", "INT-" + fila + col);
        evento.put("via", via);
        evento.put("vehiculos_contados", random.nextInt(41));
        evento.put("intervalo_segundos", 30);
        evento.put("timestamp_inicio", timestampAhora());
        evento.put("timestamp_fin", timestampAhora());
        return evento;
    }

    // genera un evento del sensor gps
    static JSONObject generarEventoGps(String fila, int col, String via) {
        boolean esNormal = random.nextInt(100) < 75;
        double velocidad;
        String nivel;
        int densidad;
        
        if (esNormal) {
            velocidad = 40.0 + random.nextDouble() * 20.0; // 40-60 km/h
            nivel = "BAJA";
            densidad = 1 + random.nextInt(18); // d < 20
        } else {
            velocidad = random.nextDouble() * 15.0; // <15 km/h
            nivel = "ALTA";
            densidad = 30 + random.nextInt(25); // d > 30
        }
        velocidad = Math.round(velocidad * 10.0) / 10.0;

        JSONObject evento = new JSONObject();
        evento.put("sensor_id", "GPS-" + fila + col + "-" + via);
        evento.put("tipo_sensor", "gps");
        evento.put("interseccion", "INT-" + fila + col);
        evento.put("via", via);
        evento.put("nivel_congestion", nivel);
        evento.put("velocidad_promedio", velocidad);
        evento.put("densidad", densidad);
        evento.put("timestamp", timestampAhora());
        return evento;
    }

    // funcion que ejecuta un sensor individual en un hilo
    static void ejecutarSensor(String tipo, String fila, int col, String via, ZContext contexto) {
        String nombre = tipo.substring(0, 3).toUpperCase() + "-" + fila + col + "-" + via;

        // simulacion de fallo por ruptura fisica apaga el 20% de los sensores si esta activo
        if (simularRuptura && random.nextInt(100) < 20) {
            System.out.println("[SIMULACION-FALLO] Sensor " + nombre + " sufrio una RUPTURA (desactivado)");
            return;
        }

        ZMQ.Socket socket = contexto.createSocket(SocketType.PUB);
        socket.connect("tcp://" + BROKER_IP + ":" + PUERTO_BROKER);

        System.out.println("[SENSOR] " + nombre + " conectado al broker");

        try { Thread.sleep(1000); } catch (InterruptedException e) { return; }  // espera inicial

        while (!Thread.currentThread().isInterrupted()) {
            JSONObject evento;
            String topico;
            if (tipo.equals("camara")) {
                evento = generarEventoCamara(fila, col, via);
                topico = "camara";
            } else if (tipo.equals("espira")) {
                evento = generarEventoEspira(fila, col, via);
                topico = "espira";
            } else {
                evento = generarEventoGps(fila, col, via);
                topico = "gps";
            }

            // marshalling json
            String mensajeJson = evento.toString();

            // simulacion de fallo por temporizacion latencia
            if (retardoMili > 0) {
                try { Thread.sleep(retardoMili); } catch (InterruptedException e) { break; }
            }

            // simulacion de fallo por omision perdida de red
            if (tasaOmision > 0 && random.nextInt(100) < tasaOmision) {
                System.out.println("[SIMULACION-FALLO] Sensor " + nombre + " -> mensaje OMITIDO (simulando perdida de canal)");
            } else {
                socket.send(topico + " " + mensajeJson);
                System.out.println("[SENSOR] " + nombre + " -> enviado | " + topico + " en " + via);
            }

            // intervalo de tiempo
            try { Thread.sleep(INTERVALO * 1000L); } catch (InterruptedException e) { break; }
        }
        socket.close();
    }

    public static void main(String[] args) {
        // parsear argumentos de simulacion de fallas y configuracion
        for (int i = 0; i < args.length; i++) {
            if (args[i].equalsIgnoreCase("-ruptura")) {
                simularRuptura = true;
            } else if (args[i].equalsIgnoreCase("-omision") && i + 1 < args.length) {
                tasaOmision = Integer.parseInt(args[i + 1]);
                i++;
            } else if (args[i].equalsIgnoreCase("-temporizacion") && i + 1 < args.length) {
                retardoMili = Integer.parseInt(args[i + 1]);
                i++;
            } else if (args[i].equalsIgnoreCase("-intervalo") && i + 1 < args.length) {
                INTERVALO = Integer.parseInt(args[i + 1]);
                i++;
            }
        }

        System.out.println("============================================================");
        System.out.println("  SENSORES DE TRAFICO - PC1 (Carrera y Calle)");
        System.out.println("============================================================");
        System.out.println("  Broker:    tcp://" + BROKER_IP + ":" + PUERTO_BROKER);
        System.out.println("  Intervalo: " + INTERVALO + " segundos");
        System.out.println("  Simulaciones de Fallos:");
        System.out.println("    - Ruptura fisica:      " + (simularRuptura ? "ACTIVADO" : "DESACTIVADO"));
        System.out.println("    - Omision de envio:    " + (tasaOmision > 0 ? tasaOmision + "% de perdida" : "DESACTIVADO"));
        System.out.println("    - Temporizacion delay: " + (retardoMili > 0 ? retardoMili + "ms" : "DESACTIVADO"));
        System.out.println("============================================================");

        ZContext contexto = new ZContext();

        // lanzar hilos para carrera y calle en cada interseccion
        for (String[] inter : INTERSECCIONES) {
            String fila = inter[0];
            int col = Integer.parseInt(inter[1]);
            for (String via : new String[]{"CARRERA", "CALLE"}) {
                for (String tipo : new String[]{"camara", "espira", "gps"}) {
                    Thread t = new Thread(() -> ejecutarSensor(tipo, fila, col, via, contexto));
                    t.setDaemon(true);
                    t.start();
                }
            }
        }

        int total = INTERSECCIONES.length * 2 * 3;
        System.out.println("\n[SENSORES] " + total + " sensores simulados iniciados");
        System.out.println("[SENSORES] Presione Ctrl+C para detener\n");

        try {
            while (true) {
                Thread.sleep(1000);
            }
        } catch (InterruptedException e) {
            System.out.println("\n[SENSORES] Cerrando...");
            contexto.close();
            System.out.println("[SENSORES] Listo.");
        }
    }
}
